File: problem_1.cpp
#include <chrono>
#include <iostream>

const int limitNumber = 1000;
const int block = 30;

bool useForSum(int n)
{
    const int numDividends = 2;
    int dividends[numDividends];
    dividends[0] = 3;
    dividends[1] = 5;
    for (int i = 0; i < numDividends; i++)
    {
       if (n % dividends[i] == 0)                   
       {
           return true;
       }
    }
    return false;
}

std::pair<int, int> sumOfBlock() 
{
    int sum = 0;
    int count = 0;
    for (int i = 0; i < block; i++) 
    {
        if (useForSum(i))
        {
            sum += i;
            count++;
        }
    }
    return std::pair<int, int> (sum, count);
}

int main() 
{
    auto startTm = std::chrono::high_resolution_clock::now();
    
    std::pair<int, int> basicBlockSum = sumOfBlock();
    long long runningSum = 0;
    int latestBlock = 0;
    // sum known blocks
    for (int i = 0; i + block < limitNumber; i+= block) 
    {
        runningSum += basicBlockSum.first;
        runningSum += basicBlockSum.second * i; 
        latestBlock = i + block;
    }
    // handle leftOver
    for (int i = latestBlock; i < limitNumber; i++) 
    {
        if (useForSum(i))                   
        {
            runningSum += i;
        }
    } 

    auto endTm = std::chrono::high_resolution_clock::now();
    auto microseconds = std::chrono::duration_cast<std::chrono::nanoseconds>(endTm-startTm);
    std::cout << "Solution: " << runningSum << std::endl;
    std::cout << "Timing: " << microseconds.count() << "ns\n";
}

File: problem_1.rs
use std::time::{Duration, Instant};

const LIMIT_NUMBER: usize = 1000;
const BLOCK_SIZE: usize = 30;

fn use_for_sum (n : usize) -> bool 
{
    return match n 
    {
        i if i % 3 == 0 => true,
        i if i % 5 == 0 => true,
        _ => false
    }
}

fn sum_of_block () -> (usize, usize)
{
    return (0usize..BLOCK_SIZE)
        .filter(|x| use_for_sum(*x))
        .map(|x| (x, 1usize))
        .fold((0usize, 0usize), |acc, x| (acc.0 + x.0, acc.1 + x.1))
}

fn main () 
{
    let start = Instant::now();
    let basic_block_sum = sum_of_block();
    let mut running_sum = 0;
    let mut latest_block = 0;
    for i in (0usize..(LIMIT_NUMBER - BLOCK_SIZE)).step_by(BLOCK_SIZE)
    {
        running_sum += basic_block_sum.0;
        running_sum += basic_block_sum.1 * i;
        latest_block = i + BLOCK_SIZE
    }
    let final_sum = (latest_block..LIMIT_NUMBER)
        .filter(|x| use_for_sum(*x))
        .fold(running_sum, |acc, x| (acc + x));

    let duration = start.elapsed();
    println!("Result: {}", final_sum);
    println!("Timing: {}", duration.as_nanos());
}